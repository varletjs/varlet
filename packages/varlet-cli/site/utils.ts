import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
