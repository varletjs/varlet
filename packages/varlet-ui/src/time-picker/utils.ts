import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
