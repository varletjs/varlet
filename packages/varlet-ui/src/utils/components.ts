import {
  createApp,
  h,
  getCurrentInstance,
  inject,
  onUnmounted,
  computed,
  provide,
  reactive,
  isVNode,
  onMounted,
  onBeforeUnmount,
  nextTick,
  ref,
  onActivated,
  onDeactivated,
} from 'vue'
import type { Component, VNode, ComputedRef, ComponentInternalInstance, Ref, ComponentPublicInstance } from 'vue'
import { isArray, removeItem } from '@varlet/shared'

export interface MountInstance {
  instance: ComponentPublicInstance
  unmount: () => void
}

export interface ChildrenCounter {
  collect(instance: ComponentInternalInstance): void

  clear(instance: ComponentInternalInstance): void

  instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
  collect(childProvider: C): void

  clear(childProvider: C): void
}

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export function mount(component: Component): MountInstance {
  const app = createApp(component)
  const host = document.createElement('div')
  document.body.appendChild(host)

  return {
    instance: app.mount(host),
    unmount() {
      app.unmount()
      document.body.removeChild(host)
    },
  }
}

export function mountInstance(
  component: any,
  props: Record<string, any> = {},
  eventListener: Record<string, any> = {}
): {
  unmountInstance: () => void
} {
  const Host = {
    setup() {
      return () =>
        h(component, {
          ...props,
          ...eventListener,
        })
    },
  }

  const { unmount } = mount(Host)
  return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
  const vNodes: VNode[] = []

  const flat = (subTree: any) => {
    if (subTree?.component) {
      flat(subTree?.component.subTree)
      return
    }

    if (Array.isArray(subTree?.children)) {
      subTree.children.forEach((child: any) => {
        if (isVNode(child)) {
          vNodes.push(child)

          flat(child)
        }
      })
    }
  }

  flat(subTree)

  return vNodes
}

export function useAtChildrenCounter(key: symbol) {
  const instances: ComponentInternalInstance[] = reactive([])
  const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  const sortInstances = () => {
    const vNodes: any[] = flatVNodes(parentInstance.subTree)

    instances.sort((a, b) => {
      return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
    })
  }

  const collect = (instance: ComponentInternalInstance) => {
    instances.push(instance)
    sortInstances()
  }

  const clear = (instance: ComponentInternalInstance) => {
    removeItem(instances, instance)
  }

  provide<ChildrenCounter>(key, {
    collect,
    clear,
    instances,
  })

  const length: ComputedRef<number> = computed(() => instances.length)

  return {
    length,
  }
}

export function useAtParentIndex(key: symbol) {
  if (!keyInProvides(key)) {
    return { index: null }
  }

  const childrenCounter: ChildrenCounter = inject<ChildrenCounter>(key) as ChildrenCounter

  const { collect, clear, instances } = childrenCounter

  const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  onMounted(() => {
    nextTick().then(() => collect(instance))
  })
  onUnmounted(() => {
    nextTick().then(() => clear(instance))
  })

  const index = computed(() => instances.indexOf(instance))

  return {
    index,
  }
}

export function useChildren<P, C>(key: symbol) {
  const childProviders: C[] = []

  const collect = (childProvider: C) => {
    childProviders.push(childProvider)
  }

  const clear = (childProvider: C) => {
    removeItem(childProviders, childProvider)
  }

  const bindChildren = (parentProvider: P) => {
    provide<P & BaseParentProvider<C>>(key, {
      collect,
      clear,
      ...parentProvider,
    })
  }

  return {
    childProviders,
    bindChildren,
  }
}

export function useParent<P, C>(key: symbol) {
  if (!keyInProvides(key)) {
    return {
      parentProvider: null,
      bindParent: null,
    }
  }

  const rawParentProvider = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>

  const { collect, clear, ...parentProvider } = rawParentProvider

  const bindParent = (childProvider: C) => {
    onMounted(() => collect(childProvider))
    onBeforeUnmount(() => clear(childProvider))
  }

  return {
    parentProvider,
    bindParent,
  }
}

export function keyInProvides(key: symbol) {
  const instance = getCurrentInstance() as any

  return key in instance.provides
}

export function useValidation() {
  const errorMessage: Ref<string> = ref('')

  const validate = async (rules: any, value: any, apis?: any): Promise<boolean> => {
    if (!isArray(rules) || !rules.length) {
      return true
    }

    const resArr = await Promise.all(rules.map((rule) => rule(value, apis)))

    return !resArr.some((res) => {
      if (res !== true) {
        errorMessage.value = String(res)
        return true
      }

      return false
    })
  }

  const resetValidation = () => {
    errorMessage.value = ''
  }

  const validateWithTrigger = async <T>(validateTrigger: T[], trigger: T, rules: any, value: any, apis?: any) => {
    if (validateTrigger.includes(trigger)) {
      ;(await validate(rules, value, apis)) && (errorMessage.value = '')
    }
  }

  return {
    errorMessage,
    validate,
    resetValidation,
    validateWithTrigger,
  }
}

export function useRouteListener(cb: () => void) {
  onMounted(() => {
    window.addEventListener('hashchange', cb)
    window.addEventListener('popstate', cb)
  })
  onUnmounted(() => {
    window.removeEventListener('hashchange', cb)
    window.removeEventListener('popstate', cb)
  })
}

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)

  onActivated(() => {
    disabled.value = false
  })

  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled,
  }
}

export function exposeApis<T = Record<string, any>>(apis: T) {
  const instance = getCurrentInstance()
  if (instance) {
    Object.assign(instance.proxy!, apis)
  }
}

type ClassName = string | undefined | null
type Classes = (ClassName | [any, ClassName, ClassName?])[]

export function createNamespace(name: string) {
  const namespace = `var-${name}`

  const createBEM = (suffix?: string): string => {
    if (!suffix) return namespace

    return suffix.startsWith('--') ? `${namespace}${suffix}` : `${namespace}__${suffix}`
  }

  const classes = (...classes: Classes): any[] => {
    return classes.map((className) => {
      if (isArray(className)) {
        const [condition, truthy, falsy = null] = className
        return condition ? truthy : falsy
      }

      return className
    })
  }

  return {
    n: createBEM,
    classes,
  }
}

export function call<F extends (...arg: any) => any, P extends Parameters<F>>(
  fn?: F | null,
  ...arg: P
): ReturnType<F> | undefined {
  if (fn) return fn(...arg)
}
