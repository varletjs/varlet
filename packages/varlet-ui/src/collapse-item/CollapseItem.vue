<template>
  <div
    :class="classes(n(), [offset && isShow, n('--active')], [disabled, n('--disable')])"
    :style="`--collapse-divider-top: ${divider ? 'var(--collapse-border-top)' : 'none'}`"
  >
    <div :class="classes(n('shadow'), formatElevation(elevation, 2))"></div>
    <div :class="n('header')" @click="toggle()">
      <div :class="n('header-title')">
        <slot name="title">{{ title }}</slot>
      </div>
      <div :class="n('header-icon')">
        <slot name="icon">
          <var-icon
            :name="icon"
            :transition="250"
            :class="
              classes(
                n('header-icon'),
                [isShow && icon === 'chevron-down', n('header-open')],
                [disabled, n('header--disable')]
              )
            "
          />
        </slot>
      </div>
    </div>
    <div
      :class="n('content')"
      v-show="showContent"
      ref="contentEl"
      @transitionend="transitionend"
      @transitionstart="start"
    >
      <div :class="n('content-wrap')">
        <slot />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, watch, computed } from 'vue'
import { nextTickFrame, requestAnimationFrame } from '../utils/elements'
import { isArray } from '@varlet/shared'
import { createNamespace, formatElevation } from '../utils/components'
import { useCollapse } from './provide'
import { props } from './props'
import VarIcon from '../icon'
import type { Ref, ComputedRef } from 'vue'
import type { CollapseItemProvider } from './provide'

const { n, classes } = createNamespace('collapse-item')

export default defineComponent({
  name: 'VarCollapseItem',
  components: {
    VarIcon,
  },
  props,
  setup(props) {
    const { index, collapse, bindCollapse } = useCollapse()

    let isInitToTrigger = true // ensure to trigger transitionend
    const contentEl: Ref<HTMLDivElement | null> = ref(null)
    const showContent: Ref<boolean> = ref(false)
    const isShow: Ref<boolean> = ref(false)
    const { active, offset, divider, elevation, updateItem } = collapse

    const name: ComputedRef<number | string | undefined> = computed(() => props.name)

    const init = (accordion: boolean, show: boolean) => {
      if (active.value === undefined || (accordion && isArray(active.value)) || show === isShow.value) return

      isShow.value = show
      toggle(true)
    }

    const toggle = (initOrAccordion?: boolean) => {
      if (props.disabled) return

      if (!initOrAccordion) {
        updateItem(props.name || index.value, !isShow.value)
      }
    }

    const openPanel = () => {
      if (!contentEl.value) return
      ;(contentEl.value as HTMLDivElement).style.height = ''
      showContent.value = true

      requestAnimationFrame(() => {
        const { offsetHeight } = contentEl.value as HTMLDivElement
        ;(contentEl.value as HTMLDivElement).style.height = 0 + 'px'

        requestAnimationFrame(() => {
          ;(contentEl.value as HTMLDivElement).style.height = offsetHeight + 'px'

          if (!isInitToTrigger) return

          nextTickFrame(() => {
            if (isInitToTrigger) transitionend()
          })
        })
      })
    }

    const start = () => {
      isInitToTrigger = false
    }

    const closePanel = () => {
      if (!contentEl.value) return

      const { offsetHeight } = contentEl.value
      contentEl.value.style.height = offsetHeight + 'px'

      requestAnimationFrame(() => {
        ;(contentEl.value as HTMLDivElement).style.height = 0 + 'px'
      })
    }

    const transitionend = () => {
      if (!isShow.value) {
        showContent.value = false
      }

      ;(contentEl.value as HTMLDivElement).style.height = ''
    }

    const collapseItemProvider: CollapseItemProvider = {
      index,
      name,
      init,
    }

    bindCollapse(collapseItemProvider)

    watch(isShow, (value) => {
      if (value) openPanel()
      else closePanel()
    })

    return {
      n,
      start,
      classes,
      showContent,
      isShow,
      offset,
      divider,
      elevation,
      toggle,
      contentEl,
      transitionend,
      formatElevation,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import './collapseItem';
</style>
