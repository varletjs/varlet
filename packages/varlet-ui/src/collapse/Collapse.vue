<template>
  <div :class="n()">
    <slot />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, nextTick, watch } from 'vue'
import { useCollapseItem } from './provide'
import { props } from './props'
import { call, createNamespace } from '../utils/components'
import { isArray } from '../utils/shared'
import type { ComputedRef } from 'vue'
import type { CollapseItemProvider } from '../collapse-item/provide'
import type { CollapseProvider } from './provide'
import type { CollapseModelValue } from './props'

const { n } = createNamespace('collapse')

export default defineComponent({
  name: 'VarCollapse',
  props,
  setup(props) {
    const { length, collapseItem, bindCollapseItem } = useCollapseItem()

    const active: ComputedRef<number | string | Array<number | string> | undefined | null> = computed(
      () => props.modelValue
    )
    const offset: ComputedRef<boolean> = computed(() => props.offset)

    const checkValue = () => {
      if (!props.accordion && !isArray(props.modelValue)) {
        console.error('[Varlet] Collapse: type of prop "modelValue" should be an Array')
        return false
      }

      if (props.accordion && isArray(props.modelValue)) {
        console.error('[Varlet] Collapse: type of prop "modelValue" should be a String or Number')
        return false
      }
      return true
    }

    const getValue = (value: number | string, isExpand: boolean): CollapseModelValue => {
      if (!checkValue()) return null
      if (isExpand) return props.accordion ? value : [...(props.modelValue as Array<string | number>), value]

      return props.accordion
        ? null
        : (props.modelValue as Array<string | number>).filter((name: string | number) => name !== value)
    }

    const updateItem = (value: number | string, isExpand: boolean) => {
      const modelValue = getValue(value, isExpand)
      call(props['onUpdate:modelValue'], modelValue)
      call(props.onChange, modelValue)
    }

    const matchName = (): Array<CollapseItemProvider> | CollapseItemProvider | undefined => {
      if (props.accordion) {
        return collapseItem.find(({ name }: CollapseItemProvider) => props.modelValue === name.value)
      }

      const filterItem = collapseItem.filter(({ name }: CollapseItemProvider) => {
        if (name.value === undefined) return false

        return (props.modelValue as Array<string | number>).includes(name.value)
      })

      return filterItem.length ? filterItem : undefined
    }

    const matchIndex = (): Array<CollapseItemProvider> | CollapseItemProvider | undefined => {
      if (props.accordion) {
        return collapseItem.find(({ index, name }: CollapseItemProvider) =>
          name.value === undefined ? props.modelValue === index.value : false
        )
      }
      return collapseItem.filter(({ index, name }: CollapseItemProvider) =>
        name.value === undefined ? (props.modelValue as Array<string | number>).includes(index.value) : false
      )
    }

    const resize = () => {
      if (!checkValue()) return

      const matchProviders: Array<CollapseItemProvider> | CollapseItemProvider | undefined = matchName() || matchIndex()

      if (
        (props.accordion && !matchProviders) ||
        (!props.accordion && !(matchProviders as Array<CollapseItemProvider>).length)
      ) {
        collapseItem.forEach((provider) => {
          provider.init(props.accordion, false)
        })
        return
      }

      collapseItem.forEach((provider) => {
        const isShow = props.accordion
          ? matchProviders === provider
          : (matchProviders as Array<CollapseItemProvider>).includes(provider)

        provider.init(props.accordion, isShow)
      })
    }

    const collapseProvider: CollapseProvider = {
      active,
      offset,
      updateItem,
    }

    bindCollapseItem(collapseProvider)

    watch(
      () => length.value,
      () => nextTick().then(resize)
    )

    watch(
      () => props.modelValue,
      () => nextTick().then(resize)
    )

    return {
      n,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';

.var-collapse {
  position: relative;
}
</style>
