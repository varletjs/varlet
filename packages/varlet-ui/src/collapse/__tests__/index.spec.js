import example from '../example'
import Collapse from '..'
import CollapseItem from '../../collapse-item'
import VarCollapse from '../Collapse'
import VarCollapseItem from '../../collapse-item/CollapseItem'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, mockConsole } from '../../utils/jest'

test('test collapse example', () => {
  const wrapper = mount(example)

  expect(wrapper.html()).toMatchSnapshot()
})

test('test collapse and collapseItem plugin', () => {
  const app = createApp({}).use(Collapse).use(CollapseItem)

  expect(app.component(Collapse.name)).toBeTruthy()
  expect(app.component(CollapseItem.name)).toBeTruthy()
})

test('test collapse modelValue and onChange', async () => {
  const template = `
    <var-collapse v-model="value" @change="changeHandle">
      <var-collapse-item title="test1" name="1">test1</var-collapse-item>
      <var-collapse-item title="test2" name="2">test2</var-collapse-item>
      <var-collapse-item title="test3" name="3">test3</var-collapse-item>
    </var-collapse>
  `

  const changeHandle = jest.fn()

  const wrapper = mount(
    {
      components: {
        [VarCollapse.name]: VarCollapse,
        [VarCollapseItem.name]: VarCollapseItem,
      },
      data() {
        return {
          value: ['2'],
        }
      },
      methods: {
        changeHandle,
      },
      template,
    },
    { attachTo: document.body }
  )

  await delay(0)

  const collapseItemList = wrapper.findAll('.var-collapse-item')

  expect(collapseItemList[1].classes()).toContain('var-collapse-item--active')

  const firstCollapseItemHeader = collapseItemList[0].find('.var-collapse-item__header')
  const secondCollapseItemHeader = collapseItemList[1].find('.var-collapse-item__header')

  expect(firstCollapseItemHeader.exists()).toBeTruthy()

  await firstCollapseItemHeader.trigger('click')

  await secondCollapseItemHeader.trigger('click')

  expect(wrapper.vm.value).toEqual(['1'])

  expect(collapseItemList[0].classes()).toContain('var-collapse-item--active')
  expect(changeHandle).toHaveBeenCalledTimes(2)
})

test('test invalid modelValue', async () => {
  const errorFn = jest.fn()
  const { mockRestore } = mockConsole('error', errorFn)
  const template = `
   <var-collapse v-model="value" :accordion="accordion">
    <var-collapse-item title="test1" name="1">test1</var-collapse-item>
    <var-collapse-item title="test2" name="2">test2</var-collapse-item>
   </var-collapse>
  `

  const wrapper = mount({
    components: {
      [VarCollapse.name]: VarCollapse,
      [VarCollapseItem.name]: VarCollapseItem,
    },
    data() {
      return {
        value: '1',
        accordion: false,
      }
    },
    template,
  })

  await delay(0)

  await wrapper.setData({
    value: ['1'],
    accordion: true,
  })

  await delay(0)

  expect(errorFn).toHaveBeenCalledTimes(2)
  mockRestore()
})

test('test collapse accordion', async () => {
  const template = `
   <var-collapse v-model="value" accordion>
    <var-collapse-item title="test1" name="1">test1</var-collapse-item>
    <var-collapse-item title="test2" name="2">test2</var-collapse-item>
   </var-collapse>
  `

  const wrapper = mount({
    components: {
      [VarCollapse.name]: VarCollapse,
      [VarCollapseItem.name]: VarCollapseItem,
    },
    data() {
      return {
        value: '2',
      }
    },
    template,
  })

  await delay(0)

  await wrapper.find('.var-collapse-item__header').trigger('click')

  expect(wrapper.vm.value).toBe('1')
})

test('test collapseItem disabled', async () => {
  const template = `
    <var-collapse v-model="value">
      <var-collapse-item title="test1" disabled name="1">test1</var-collapse-item>
      <var-collapse-item title="test2" name="2">test2</var-collapse-item>
      <var-collapse-item title="test3" name="3">test3</var-collapse-item>
    </var-collapse>
  `

  const wrapper = mount(
    {
      components: {
        [VarCollapse.name]: VarCollapse,
        [VarCollapseItem.name]: VarCollapseItem,
      },
      data() {
        return {
          value: ['2'],
        }
      },
      template,
    },
    { attachTo: document.body }
  )

  await delay(0)

  const collapseItemList = wrapper.findAll('.var-collapse-item')

  expect(collapseItemList[0].classes()).toContain('var-collapse-item--disable')

  await collapseItemList[0].find('.var-collapse-item__header').trigger('click')

  expect(wrapper.vm.value.toString()).toBe('2')
})

test('test collapse offset and collapse icon', () => {
  const template = `
    <var-collapse v-model="value" :offset="false">
      <var-collapse-item title="test1" icon="fire">test1</var-collapse-item>
      <var-collapse-item title="test2" icon="fire">test2</var-collapse-item>
    </var-collapse>
  `

  const wrapper = mount(
    {
      components: {
        [VarCollapse.name]: VarCollapse,
        [VarCollapseItem.name]: VarCollapseItem,
      },
      data() {
        return {
          value: [2],
        }
      },
      template,
    },
    { attachTo: document.body }
  )

  expect(wrapper.html()).toMatchSnapshot()
})
