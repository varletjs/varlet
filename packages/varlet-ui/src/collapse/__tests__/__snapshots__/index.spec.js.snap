// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test collapse example 1`] = `
"<div class=\\"collapse-demo\\">
  <div>
    <div class=\\"app-type\\">基本使用</div>
    <div class=\\"var-collapse\\">
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">隐藏边距</div>
    <div class=\\"var-collapse\\">
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">手风琴模式</div>
    <div class=\\"var-collapse\\">
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">禁用</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--2\\" style=\\"margin-bottom: 8px;\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">禁用</div>
    </button>
    <div class=\\"var-collapse\\">
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">自定义内容</div>
    <div class=\\"var-collapse\\">
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">这是标题</div>
          <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-account-circle var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\">文本</div>
        </div>
      </div>
      <div class=\\"var-collapse-item\\">
        <div class=\\"var-collapse-item__header\\">
          <div class=\\"var-collapse-item__header-title\\">这是标题</div>
          <div class=\\"var-collapse-item__header-icon\\">^_^</div>
        </div>
        <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
          <div class=\\"var-collapse-item__content-wrap\\"> 这是内容</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`test collapse offset and collapse icon 1`] = `
"<div class=\\"var-collapse\\">
  <div class=\\"var-collapse-item\\">
    <div class=\\"var-collapse-item__header\\">
      <div class=\\"var-collapse-item__header-title\\">test1</div>
      <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-fire var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
    </div>
    <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
      <div class=\\"var-collapse-item__content-wrap\\">test1</div>
    </div>
  </div>
  <div class=\\"var-collapse-item\\">
    <div class=\\"var-collapse-item__header\\">
      <div class=\\"var-collapse-item__header-title\\">test2</div>
      <div class=\\"var-collapse-item__header-icon\\"><i class=\\"var-icon var-icon--set var-icon-fire var-collapse-item__header-icon\\" style=\\"transition: transform 250ms;\\"></i></div>
    </div>
    <div class=\\"var-collapse-item__content\\" style=\\"display: none;\\">
      <div class=\\"var-collapse-item__content-wrap\\">test2</div>
    </div>
  </div>
</div>"
`;
